packages/eslint-plugin/src/rules/await-thenable.ts

```diff
@@ -62,6 +72,55 @@ export default createRule({
           });
         }
       },
+
+      ForOfStatement(node): void {
+        if (!node.await) {
+          return;
+        }
+
+        const type = services.getTypeAtLocation(node.right);
+        if (isTypeAnyType(type) || isTypeUnknownType(type)) {
+          return;
+        }
+
+        const asyncIteratorSymbol = tsutils.getWellKnownSymbolPropertyOfType(
+          type,
+          'asyncIterator',
+          checker,
+        );
+
+        // if there is an async iterator symbol, but it doesn't have the correct
+        // shape, TS will report a type error, so we only need to check if the
+        // symbol exists.
+        if (asyncIteratorSymbol == null) {
+          context.report({
+            loc: getForStatementHeadLoc(context.sourceCode, node),
+            messageId: 'forAwaitOfNonThenable',
+            suggest: [
+              // This suggestion causes broken code for sync iterables of promises, since
+              // the loop variable will not be awaited.
+              //
+              // Ideally, if the iterable yields promises, we would offer a suggestion to
```

Comment:
bleh - I guess this isn't always a good idea, particularly with possibly infinite iterables. But, then again, why would you have an infinite sync iterable of promises? Thinking out loud.